### PR TITLE
Add a `get_leaf_nodes` utility function

### LINKED ISSUE #661: Add the ability to determine what leaf nodes a test point falls in.
First, incorporate find_leaf_node function into the package.
```
find_leaf_node <- function(tree, sample) {
  node <- 1
  while (TRUE) {
    if (tree$nodes[[node]]$is_leaf) {
      break()
    }
    split_var <- tree$nodes[[node]]$split_variable
    split_value <- tree$nodes[[node]]$split_value
    if (sample[split_var] <= split_value) {
      node <- tree$nodes[[node]]$left_child
    } else {
      node <- tree$nodes[[node]]$right_child
    }
  }
  node
}
```

Second, find_leaf_node gives external numbering but s.forest$`_leaf_samples`[[1]] is internal.

### PR DESCRIPTION
**Pull request by erikcs**
As mentioned by @susanathey, a recurring question is how to retrieve the leaf nodes a test sample falls into in a given tree.

This PR adds a simple utility function `get_leaf_nodes(tree, test.sample)` which for a given GRF tree object, returns the leaf node number for the points in `test.sample`. The docstring refers to `get_sample_weights`, emphasizing this function is not indented for statistical estimation tasks...

It is cumbersome to call into C++ to do this and it is done in R here instead for simplicity, I think the speed is acceptable considering this is not a "core" workflow.

```R
n <- 5000
p <- 10
X <- matrix(rnorm(n * p), n, p)
Y <- X[, 1] * rnorm(n)
r.forest <- regression_forest(X, Y, num.trees = 1)
tree <- get_tree(r.forest, 1)
nt <- 20000
Xt <- matrix(rnorm(nt * p), nt, p)
system.time(n <- get_leaf_node(tree, Xt))
user  system elapsed 
1.912   0.005   1.920 
```

Closes #661. 

(Will be work to merge into the "XPtr serialization draft [On hold]" PR)

### CODE REVIEW

File: r-package/grf/R/analysis_tools.R [hunk 1]
```diff
@@ -213,6 +213,75 @@ get_sample_weights <- function(forest, newdata = NULL, num.threads = NULL) {
...
+  if (!("grf_tree" %in% class(tree))) {
+    stop("get_leaf_node is only implemented for class `grf_tree`")
+  }
+  validate_X(newdata, allow.na = TRUE)
+  num.samples <- nrow(newdata)
+  num.features <- ncol(newdata)
+  leaf.nodes <- rep(0, num.samples)
+
+  for (i in 1:num.samples) {
+    node <- 1
```
Comment (halflearned) [inline r-package/grf/R/analysis_tools.R:252]:
```suggestion
    # Start at root node
    node <- 1
```

File: r-package/grf/R/analysis_tools.R [hunk 2]
```diff
@@ -213,6 +213,75 @@ get_sample_weights <- function(forest, newdata = NULL, num.threads = NULL) {
   }
 }
 
+#' Find the leaf node for a test sample.
+#'
+#' Given a GRF tree object, compute the leaf node a test sample falls into. The nodes in a GRF tree
+#' is numbered according to breadth first, the returned numbers will be the leaf integer according
```
Comment (halflearned) [inline r-package/grf/R/analysis_tools.R:219]:
```suggestion
#' are numbered breadth first, and the returned numbers will be the leaf integer according
```

**Review by halflearned — COMMENTED** [review]
Thanks, @erikcs! A question: are nodes 0- or 1-indexed? The output number suggests 1-indexed, but if I plot a tree and hover my mouse above it, the tooltip seems to be 0-indexed. See screenshots:
tooltip

**Comment by erikcs** [comment]
Yes, the Graphviz plot numbers them from 0, but the R indexes are starting at 1.

**Review by halflearned — APPROVED** [review]

**Pull request merged**